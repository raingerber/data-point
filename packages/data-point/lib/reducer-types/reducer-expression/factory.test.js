/* eslint-env jest */
'use strict'

const _ = require('lodash')
const TransformFactory = require('./factory')

describe('isValid', () => {
  test('It should return true if type is valid', () => {
    expect(TransformFactory.isValid('string')).toEqual(true)
    expect(TransformFactory.isValid(() => 1)).toEqual(true)
    expect(TransformFactory.isValid([])).toEqual(true)
    expect(TransformFactory.isValid({})).toEqual(true)
  })

  test('should return false if invalid', () => {
    expect(TransformFactory.isValid(1)).toEqual(false)
    expect(TransformFactory.isValid(true)).toEqual(false)
  })
})

describe('validate', () => {
  test('It should throw error if transform is invalid', () => {
    expect(() => {
      TransformFactory.validate(true)
    }).toThrowErrorMatchingSnapshot()
  })

  test('it should return true if valid', () => {
    expect(TransformFactory.validate('$.')).toEqual(true)
  })
})

describe('TransformFactory#create', () => {
  test('TransformFactory#create default', () => {
    const result = TransformFactory.create()

    expect(result).toBeInstanceOf(TransformFactory.ReducerExpression)
    expect(result.context).toBeUndefined()
    expect(result.reducers).toHaveLength(0)
  })

  test('TransformFactory#create only path', () => {
    const result = TransformFactory.create('$foo.bar')
    expect(result.reducers).toHaveLength(1)

    const reducer = _.first(result.reducers)
    expect(reducer.type).toBe('ReducerPath')
    expect(reducer.name).toBe('foo.bar')
    expect(reducer.asCollection).toBe(false)
  })

  test('TransformFactory#create path as collection', () => {
    const result = TransformFactory.create('$foo.bar[]')
    expect(result.reducers).toHaveLength(1)

    const reducer = _.first(result.reducers)
    expect(reducer.type).toBe('ReducerPath')
    expect(reducer.name).toBe('foo.bar')
    expect(reducer.asCollection).toBe(true)
  })

  test('TransformFactory#create context with reducers', () => {
    const result = TransformFactory.create([
      '$foo.bar | reducer:add',
      () => true
    ])

    expect(result.reducers).toHaveLength(3)
    expect(result.reducers[0].type).toBe('ReducerPath')
    expect(result.reducers[1].type).toBe('ReducerEntity')
    expect(result.reducers[2].type).toBe('ReducerFunction')
  })

  test('TransformFactory#create context with reducers', () => {
    const result = TransformFactory.create([
      '$foo.bar',
      'reducer:add',
      () => true
    ])

    expect(result.reducers).toHaveLength(3)
    expect(result.reducers[0].type).toBe('ReducerPath')
    expect(result.reducers[1].type).toBe('ReducerEntity')
    expect(result.reducers[2].type).toBe('ReducerFunction')
  })
})

describe('TransformFactory#create', () => {
  test('TransformFactory#create transfrom empty array', () => {
    const result = TransformFactory.create([])
    expect(result.reducers).toHaveLength(0)
  })

  test('TransformFactory#create transfrom with single function', () => {
    const result = TransformFactory.create((acc, done) => done(null, acc.value))
    expect(result.reducers).toHaveLength(1)
    expect(result.reducers[0].type).toBe('ReducerFunction')
  })

  test('TransformFactory#create transfrom with string', () => {
    const result = TransformFactory.create(['$foo', '$bar'])
    expect(result.reducers).toHaveLength(2)
    expect(result.reducers[0].type).toBe('ReducerPath')
    expect(result.reducers[1].type).toBe('ReducerPath')
  })

  test('TransformFactory#create transfrom with grouped reducers and single reducers', () => {
    const result = TransformFactory.create(['$foo | $bar', '$baz'])
    expect(result.reducers).toHaveLength(3)
    expect(result.reducers[0].type).toBe('ReducerPath')
    expect(result.reducers[1].type).toBe('ReducerPath')
    expect(result.reducers[2].type).toBe('ReducerPath')
  })

  test('TransformFactory#create errors if invalid reducer type', () => {
    const result = _.attempt(TransformFactory.create, 1)

    expect(result).toBeInstanceOf(Error)
  })

  test('TransformFactory#create transfrom from array', () => {
    const result = TransformFactory.create([
      '$foo.bar',
      'reducer:add | $foo.bar.zeta',
      (acc, done) => done(null, acc.value)
    ])

    expect(result.reducers).toHaveLength(4)
    expect(result.reducers[0].type).toBe('ReducerPath')
    expect(result.reducers[1].type).toBe('ReducerEntity')
    expect(result.reducers[2].type).toBe('ReducerPath')
    expect(result.reducers[3].type).toBe('ReducerFunction')
  })
})
