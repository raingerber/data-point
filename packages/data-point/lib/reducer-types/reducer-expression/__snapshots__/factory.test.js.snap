// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate It should throw error if transform is invalid 1`] = `
"Could not parse a ReducerExpression. The ReducerExpression:
 true
is not using a valid type, try using an Array, String, Object, or Function.
More info: https://github.com/ViacomInc/data-point/tree/master/packages/data-point#reducer-expression
"
`;
